File: analysis/macros/ZeroBias/PU_factor.cc
#include <vector>
#include <iostream>
#include <TTree.h>
#include <TString.h>
#include <TMath.h>
#include <TH1.h>
#include <algorithm>
#include <nlohmann/json.hpp>

const int nbins = 90;

int PU_factor(TString era){

    TFile *datafile = TFile::Open("/pnfs/iihe/cms/store/user/kakang/IPres/analysis/tuples/ZeroBias/all_skimmed_2022_data_"+era+".root");
    TTree *datatree = (TTree*)datafile->Get("mytree");
    TFile *mcfile = TFile::Open("/pnfs/iihe/cms/store/user/kakang/IPres/analysis/tuples/ZeroBias/all_skimmed_2022_mc_"+era+".root");
    TTree *mctree = (TTree*)mcfile->Get("mytree");

    int dataev_nPV;
    datatree->SetBranchAddress("ev_nPV", &dataev_nPV);
    int mcev_nPV;
    mctree->SetBranchAddress("ev_nPV", &mcev_nPV);

    TH1F * h_data = new TH1F("h_data", "", nbins, 0.5, nbins+0.5);
    TH1F * h_mc = new TH1F("h_mc", "", nbins, 0.5, nbins+0.5);

    datatree->Project("h_data", "ev_nPV");
    mctree->Project("h_mc", "ev_nPV");

    h_data->Scale(1./h_data->Integral());
    h_mc->Scale(1./h_mc->Integral());

    TH1F * h_ratio = (TH1F*)h_data->Clone("h_ratio");
    h_ratio->Divide(h_mc);
    
    std::vector<float> dataPU_fac;
    std::vector<float> mcPU_fac;

    for(int i=0; i<5; i++){
        if(era.Contains("preEE")){ 
            dataPU_fac.push_back(0);
            mcPU_fac.push_back(0);
        }
        else if(era.Contains("postEE")){
            dataPU_fac.push_back(1);
            mcPU_fac.push_back(h_ratio->GetBinContent(i+1));
        }
    }

    for(int i=5; i<nbins; i++){
        dataPU_fac.push_back(1);
        mcPU_fac.push_back(h_ratio->GetBinContent(i+1));
    }

    TFile *dataoutfile = new TFile("/pnfs/iihe/cms/store/user/kakang/IPres/analysis/tuples/ZeroBias/all_skimmed_2022_data_"+era+"_corr.root", "RECREATE");
    TTree *dataouttree = datatree->CloneTree(0);
    float datafactor;
    dataouttree->Branch("PU_factor", &datafactor);
    for(int i=0; i<datatree->GetEntries(); i++){
        datatree->GetEntry(i);
        datafactor = dataPU_fac[dataev_nPV-1];
        dataouttree->Fill();
    }
    dataouttree->Write();
	delete dataouttree;
	dataoutfile->Close();

    TFile *mcoutfile = new TFile("/pnfs/iihe/cms/store/user/kakang/IPres/analysis/tuples/ZeroBias/all_skimmed_2022_mc_"+era+"_corr.root", "RECREATE");
    TTree *mcouttree = mctree->CloneTree(0);
    float mcfactor;
    mcouttree->Branch("PU_factor", &mcfactor);
    for(int i=0; i<mctree->GetEntries(); i++){
        mctree->GetEntry(i);
        mcfactor = mcPU_fac[mcev_nPV-1];
        mcouttree->Fill();
    }
    mcouttree->Write();
	delete mcouttree;
	mcoutfile->Close();

    return 0;
}

File: analysis/macros/ZeroBias/PU_factor.py
import sys
import ROOT
from ROOT import RDataFrame, RDF, gInterpreter
import json
import numpy as np

nbins = 90

with open("tuplelist.json") as f:
    cfg = json.load(f)


def build_globs_by_period(cfg: dict, period: str):
    if period not in cfg:
        raise KeyError(
            f"period '{period}' not found in config keys: {list(cfg.keys())}"
        )

    section = cfg[period]
    data_entries = section.get("data", []) or []
    mc_entries = section.get("mc", []) or []

    data_files = [
        f"/eos/home-k/kakang/IPres/analysis/ZeroBias/tuples/{period}/{era['dataset']}/skimmed_*.root"
        for era in data_entries
        if "dataset" in era
    ]
    mc_files = [
        f"/eos/home-k/kakang/IPres/analysis/ZeroBias/tuples/{period}/{era['dataset']}/skimmed_*.root"
        for era in mc_entries
        if "dataset" in era
    ]

    return data_files, mc_files


def make_hist(df):
    model = ROOT.RDF.TH1DModel("h", "", nbins, 0.5, nbins + 0.5)
    h = df.Histo1D(model, "ev_nPV").GetValue()
    if h.Integral() > 0:
        h.Scale(1.0 / h.Integral())
    return h


def build_factors(h_ratio):
    dataPU_fac, mcPU_fac = [], []
    for i in range(nbins):
        dataPU_fac.append(1.0)
        mcPU_fac.append(h_ratio.GetBinContent(i + 1))
    return dataPU_fac, mcPU_fac


def declare_lookup_func(name, factors):
    init = ", ".join(f"{x:.7g}" for x in factors)
    code = f"""
    float {name}(int npv){{
        static const float lut[{nbins}] = {{{init}}};
        return lut[npv-1];
    }}
    """
    ROOT.gInterpreter.Declare(code)

def PU_factor(period: str) -> int:

    data_in, mc_in = build_globs_by_period(cfg, period)

    df_data = ROOT.RDataFrame("mytree", data_in)
    df_mc = ROOT.RDataFrame("mytree", mc_in)

    h_data = make_hist(df_data)
    h_mc = make_hist(df_mc)

    h_ratio = h_data.Clone("h_ratio")
    h_ratio.Divide(h_mc)

    dataPU_fac, mcPU_fac = build_factors(h_ratio)

    declare_lookup_func("compute_factor_data", dataPU_fac)
    declare_lookup_func("compute_factor_mc", mcPU_fac)

    out_data = (
        f"/eos/home-k/kakang/IPres/analysis/ZeroBias/tuples/{period}/data_corr.root"
    )
    out_mc = f"/eos/home-k/kakang/IPres/analysis/ZeroBias/tuples/{period}/mc_corr.root"

    df_data.Define("PU_factor", "compute_factor_data(ev_nPV)").Snapshot(
        "mytree", out_data
    )
    df_mc.Define("PU_factor", "compute_factor_mc(ev_nPV)").Snapshot("mytree", out_mc)

    return 0


if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: python PU_factor.py <period>  # e.g. 2022preEE or 2022postEE")
        sys.exit(1)
    sys.exit(PU_factor(sys.argv[1]))
